src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -251,7 +249,15 @@ private Type EmitIListCoercion<T>()
             EmitOpCode(OpCodes.Ldloc, counterIndex);
 
             // Load array to be coerced.
-            EmitOpCode(OpCodes.Ldloc, localVarIndex);
+            var currentVarIndex = localVarIndex;
+            if (arg is IdentifierNode ident)
+            {
+                currentVarIndex = variables[ident.Value].Item1;
+            }
```

Comment:
If arg isn't an Identifier, how will this work? There will be no dictionary entry, nor will `localVarIndex` be correct.

Reply:
yes, thats why I've added the TODO below for the other case where `emitExpressionListNode` is used.

Reply:
ok, looks good.

Reply:
So, the reason why you're seeing a regression with that test case in your PR is that the current `localVarIndex` is not pointing the array to be coerced but to something else that has been added to the top of the stack?

Reply:
it's because the `localIndexVar` variable is incorrect when we get to the `emitMSIL` phase - it's already been incremented in the `GatherTypeInfo` phase.

this change avoids both issues though because we lookup the associated index for the specific variable, not just whatever happens to be the last variable we declared.

Reply:
I am still working through similar issues with the operators tests though, so maybe more fixes will be required here.

Reply:
So with this fix, can you reset the localVarIndex after the GatherTypeInfo compile pass?

Reply:
With the current implementation it will not help - let me explain.
1. in this PR - in the `GatherTypeInfo` phase - we increment `localvarIndex` and declare variables in the variables dictionary.
2. If we declare / increment those variables again, we'll throw exceptions, there is an invariant in the current code that variables should not be redefined.
3. If we do a reset between phases, then the local var index will just always be -1 during`emitIL` phase.

... I guess we could just increment the `localVarIndex` during `emitIL` phase so at least it is correct, but not touch the `variables` dictionary.

do you want me to investigate that so it's a bit safer to use the `localVarIndex` during `emitIL` or alternatively we could add some documentation to this field?

Reply:
I was under the impression that variables are being declared and added to the dictionary in both phases, and that `localVarIndex` continues incrementing from where it left off in the `emitIL `phase. If I understand correctly, it sounds like now, no variables are declared in the `emitIL` phase any longer and the `variables `dictionary and `localVarIndex` are up-to-date at the end of the `GatherTypeInfo` phase. If that's the case, it would be unsafe to use `localVarIndex` in the `emitIL `phase altogether. We would need to query the `variables` dictionary using the identifier name in all places instead as you've done here.

Reply:
yes, in the current PR is it not safe to use the `localVarIndex` in the `emitIL` phase at all. 
I see a few alternatives to make this less of an issue.

1. try to reset and re-increment the localVarIndex in the `emitIL` phase... it's likely still going to lead to bugs like we saw today though when larger programs start declaring multiple variables. (I think the use of the index from the `variables` dictionary is a safer approach in general).
2. document that the localVarIndex should not be used in `emitIL`
3. wrap localVarIndex in a getter and throw when trying to access in `emitIL`

Reply:
I don't like option 1. I'm fine with option 2. 
I wonder if it's even necessary to continue keeping `localVarIndex` as a class member. I guess I'll need to review this PR again.